Flac header extractor

noahmacaskill opened the pull request:
Extracted metadata from flac header.

noahmacaskill commented:
Extracted what I thought were all the important pieces of metadata from the flac header. Had to change the metadata command test since with this change, it was failing at trying to open a filestream from our mock file system. Instead just fed it a path to some of the Fixtures files.

atruskie reviewed (changes requested):
Wow, excellent changes. I'm impressed.

- we update the docs/metadata.md with ✅for the metadata now supported by this change
- it would be great to see more fixtures tested
- I'm a bit dubious of the durations you entered into the CSV fixtures file. Make sure they're exact numbers and not floating point approximations. We probably want to change the fixture model to consider duration a `decimal` as well
- It would be good if you tested the methods you added to the `Flac` class as well

Really nice though!

noahmacaskill commented on src/MetadataUtility/Metadata/FrontierLabs/FlacHeaderExtractor.cs:
```diff
@@ -7,28 +7,45 @@ namespace MetadataUtility.Metadata.FrontierLabs
-            var result = !information.IsFlacFile();
+            var result = information.IsFlacFile();
 
             return ValueTask.FromResult(result);
```
Initially made an assertion for this, but it broke the MetadataCommand test (since the empty files don't have enough bytes for a header). Instead I added this logic into the CanProcessAsync, changing IsFlacFile to IsValidFlacFile, does that work?

noahmacaskill replied:
Also, I verified the stream was at least [42 bytes](https://github.com/QutEcoacoustics/emu/pull/121/files#:~:text=buffer.StartsWith(FlacMagicNumber)%20%26%26-,stream.Length%20%3E%2042,-%3B), which includes everything up to the end of the [metadata black stream info](https://xiph.org/flac/format.html#metadata_block_streaminfo).

noahmacaskill commented on test/MetadataUtility.Tests/Metadata/FrontierLabs/FlacHeaderExtractorTests.cs:
```diff
@@ -48,13 +46,20 @@ public async void CanProcessFilesWorks(FixtureModel model)
+            var recording = new Recording();
 
-            recording.DurationSeconds?.TotalSeconds.Should().Be(model.DurationSeconds);
+            if (await this.subject.CanProcessAsync(model.ToTargetInformation(this.RealFileSystem)))
```
My thinking here was, some of the fixture models (namely the wav header one) isn't going to pass CanProcessAsync, so we don't want it running ProcessAsync. I've changed the if statement to `if (model.IsFlac)`, and used the same if statement in my FlacTests file as well. Let me know your thoughts.

noahmacaskill replied:
I should also note that I changed the MimeType value for the PreallocatedHeader fixtures csv to audio/wav instead of audio/flac, so that it returns false for model.IsFlac.

atruskie replied:
So yeah, these cases will be a little tricky.

If some of our fixtures are not appropriate for some tests, we should filter them out. That is what the `IsFlac` property was for and you correctly keyed into that. It might not be the best property to filter on though.

Then we have to decide what happens for faulty files. For CanProcessAsync the goal is to never have it run if there is file it can't process. That won't always be true, so we'll code defensively anyway, 

On reflection I think your original approach is pretty decent, but we'll tweak it a little bit. Something like:

```
if (!  if (await this.subject.CanProcessAsync(model.ToTargetInformation(this.RealFileSystem)))) {
   Assert.Ignore("Test filtered")
}
```

atruskie replied:
Another approach might be to keep a list of which extractors we expect any fixture to run on. That could be the best approach in the long run.

So in the fixtures CSV we'd have a column for each fixture named `CanProccess`  and `Process`. The values for those columns will be a semi-colon delimited list of which extractors we expect to be tested for each file.

It might be more work in the short run, but it the long run it will allow us to be very explicit about what we test.

noahmacaskill commented:
For duration, the nodatime Duration seems to only work for double precision, not decimal. I changed duration to just be a simple decimal type instead of a Duration object, and made it as precise as I could get it in the fixtures csv. Open to suggestions here if you have any ideas.

Last thing to do is include more fixtures to test in the csv, will get on that early next week, does it make sense to just test every fixture we have here?

atruskie commented on src/MetadataUtility/Metadata/Predicates.cs:
```diff
@@ -13,7 +13,7 @@ public static class Predicates
         target => !string.IsNullOrWhiteSpace(target.FileSystem.Path.GetFileName(target.Path));
 
     public static readonly Func<TargetInformation, bool> IsFlacFile =
-        target => Flac.IsFlacFile(target.FileStream).IfFail(false);
+        target => Flac.IsValidFlacFile(target.FileStream).IfFail(false);
```
If you keep the name, update the wrapper extension method to reflect what is is actually doing (`IsFlacFile` -> `IsValidFlacFile`)

But (see previous comments) I see value in granular predicates.

atruskie commented on src/MetadataUtility/Metadata/TargetInformation.cs:
```diff
@@ -55,11 +55,11 @@ public TargetInformation(IFileSystem fileSystem)
             get
             {
-                return this.stream ?? (FileStream)this.FileSystem.File.Open(this.Path, FileMode.Open, FileAccess.Read, FileShare.Read);
+                return this.stream ?? (Stream)this.FileSystem.File.Open(this.Path, FileMode.Open, FileAccess.Read, FileShare.Read);
```
This cast should be unnecessary?

atruskie reviewed (approved):
I've made a few minor suggestions, but I think we're close to merging. Great work again.

We can add more test files later if needed. Let me know when you think you're done and if tests are passing we'll merge.

noahmacaskill commented:
Think I covered all the bases in terms of the comments.

For the precision of duration in the fixtures CSV, I've turned this into a fraction (samples/sample rate) to keep it exact, and added functionality to the csv serializer to have this be read in as rational number.

Divided the IsValidFlacFile into the two functions, and added some additional tests for those as well.

Lastly I added the Process column to the csv like you suggested (added csv serializer functionality to read in lists as well).

Feeling like it probably makes sense to test my serialization converters too, maybe something similar to [NodatimeSerializerTests?](https://github.com/QutEcoacoustics/emu/blob/master/test/MetadataUtility.Tests/Serialization/NodatimeSerializersTests.cs), though that's testing both writing to and reading from the serializer (I could implement a ConvertToString if needed to test).

Let me know what you think.

atruskie reviewed (approved):
Ok, just add the changes to docs/metadata and we're good to go!

noahmacaskill commented:
Updated Metadata.md, in my mind these changes aren't vendor specific, but should work for any FLAC file (same will be true for WAV)? Reorganized the document with that in mind.

atruskie commented:
Excellent work!

Note I've squashed your commits: take care to switch back to master and delete your local copy of your branch
